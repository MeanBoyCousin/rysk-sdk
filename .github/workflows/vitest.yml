name: Vitest

on:
  pull_request:
    branches: [master]

  workflow_dispatch:

concurrency:
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: ${{ github.ref != 'refs/heads/main' }}

jobs:
  test:
    name: Test
    permissions:
      actions: write
      contents: read
    runs-on: ubuntu-latest
    timeout-minutes: 2

    steps:
      - name: Checkout
        uses: actions/checkout@v4

      - name: Setup Node.js
        uses: actions/setup-node@v4
        with:
          cache: 'yarn'
          node-version: '20.x'

      - name: Install packages
        run: yarn

      - name: Run tests
        run: yarn test --silent
