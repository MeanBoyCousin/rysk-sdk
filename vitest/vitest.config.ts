import { resolve } from 'path'

import { configDefaults, defineConfig } from 'vitest/config'

const config = defineConfig({
  resolve: {
    alias: [
      { find: 'src', replacement: resolve(__dirname, '../', './src') },
      { find: 'vitest/utils', replacement: resolve(__dirname, '../', './vitest/utils.ts') },
    ],
  },
  root: './src',
  test: {
    coverage: {
      all: true,
      exclude: [
        '__mocks__/',
        '__tests__/',
        'types/',
        '**/types.ts',
        '**/*.d.ts',
        'constants/**',
        'ABI/**',
        'sandbox.ts',
      ],
      ignoreEmptyLines: true,
      reporter: ['html', 'lcov'],
      reportsDirectory: '../reports/vitest',
      thresholds: {
        lines: 100,
        branches: 100,
        functions: 100,
        statements: 100,
      },
      watermarks: {
        statements: [80, 100],
        functions: [80, 100],
        branches: [80, 100],
        lines: [80, 100],
      },
    },
    environment: 'node',
    globals: true,
    setupFiles: '../vitest/vitest.setup.ts',
    fakeTimers: {
      toFake: [...(configDefaults.fakeTimers.toFake ?? []), 'performance'],
    },
  },
})

export default config
