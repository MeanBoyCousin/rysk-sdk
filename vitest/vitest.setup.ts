import { vi } from 'vitest'
import createFetchMock from 'vitest-fetch-mock'

vi.spyOn(console, 'error').mockImplementation(() => {})
vi.spyOn(console, 'log').mockImplementation(() => {})

const fetchMocker = createFetchMock(vi)
fetchMocker.enableMocks()

Object.defineProperty(BigInt.prototype, 'toJSON', {
  get() {
    'use strict'
    return () => String(this)
  },
})

vi.stubEnv('CIAO_MAINNET_ADDRESS', '0x71728FDDF90233cc35D61bec7858d7c42A310ACe')
vi.stubEnv('CIAO_TESTNET_ADDRESS', '0x71728FDDF90233cc35D61bec7858d7c42A310ACe')
vi.stubEnv('ORDER_DISPATCH_MAINNET_ADDRESS', '0x27809a3Bd3cf44d855f1BE668bFD16D34bcE157C')
vi.stubEnv('ORDER_DISPATCH_TESTNET_ADDRESS', '0x27809a3Bd3cf44d855f1BE668bFD16D34bcE157C')
