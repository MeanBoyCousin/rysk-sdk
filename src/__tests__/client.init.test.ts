import { beforeEach, describe, expect, it, vi } from 'vitest'

import RyskSDK from 'src'
import { Environment } from 'src/enums'
import { privateKey } from 'vitest/utils'

describe('The RyskSDK', () => {
  beforeEach(() => {
    fetchMock.resetMocks()
  })

  it('should initialise correctly with only required params', () => {
    const Client = new RyskSDK(privateKey)
    Client.publicClient.uid = 'test-uid'
    expect(Client).toMatchSnapshot()
  })

  it('should initialise correctly with all config parameters passed', () => {
    const Client = new RyskSDK(privateKey, {
      debug: true,
      environment: Environment.TESTNET,
      rpc: 'https://test-rpc.quiknode.pro',
      subAccountId: 2,
    })
    Client.publicClient.uid = 'test-uid'
    expect(Client).toMatchSnapshot()
  })

  it('should POST the referral code if running on mainnet', async () => {
    fetchMock.mockResponse(JSON.stringify({ success: true }))

    new RyskSDK(privateKey, { environment: Environment.MAINNET })

    await vi.waitFor(() => {
      if (!fetchMock.mock.calls.length) throw new Error('Awaiting call...')
    })

    expect(fetchMock.mock.lastCall).toMatchInlineSnapshot(`
      [
        "https://api.rysk.finance/v1/vault/referral",
        {
          "body": "{"account":"0xb47B0b1e44B932Ae9Bb01817E7010A553A965Ea8","code":"kickflip"}",
          "method": "POST",
        },
      ]
    `)
  })

  it('should handle referral failures gracefully', async () => {
    fetchMock.mockReject(new Error('An unknown error occurred'))

    const Client = new RyskSDK(privateKey, { debug: true, environment: Environment.MAINNET })

    await vi.waitFor(() => {
      if (Client.logs.length !== 3) throw new Error('Awaiting call...')
    })

    expect(Client.logs[2]).toEqual({ msg: 'Call failed, ignoring.' })
  })
})
