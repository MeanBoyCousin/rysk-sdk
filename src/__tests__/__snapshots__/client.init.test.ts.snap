// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The RyskSDK > should initialise correctly with all config parameters passed 1`] = `
RyskSDK {
  "account": {
    "address": "0xb47B0b1e44B932Ae9Bb01817E7010A553A965Ea8",
    "experimental_signAuthorization": [Function],
    "nonceManager": undefined,
    "publicKey": "0x0422eea015a8664523fce2eed5de514ad98a29724755c488f3d11d913d5671c881bc2d7b1e0d17389604ef141a96830b14917b034b7621579d1159d694dd231798",
    "sign": [Function],
    "signMessage": [Function],
    "signTransaction": [Function],
    "signTypedData": [Function],
    "source": "privateKey",
    "type": "local",
  },
  "calculateMarginRequirement": [Function],
  "cancelAndReplaceOrder": [Function],
  "cancelOpenOrdersForProduct": [Function],
  "cancelOrder": [Function],
  "cancelOrders": [Function],
  "chain": 421614,
  "ciaoAddress": "0x71728FDDF90233cc35D61bec7858d7c42A310ACe",
  "deposit": [Function],
  "domain": {
    "chainId": 421614n,
    "name": "rysk",
    "verifyingContract": "0x27809a3Bd3cf44d855f1BE668bFD16D34bcE157C",
    "version": "0.0.0",
  },
  "environment": "testnet",
  "getAccountHealth": [Function],
  "getKlines": [Function],
  "getOrderBook": [Function],
  "getProduct": [Function],
  "getProducts": [Function],
  "getServerTime": [Function],
  "getTickers": [Function],
  "getTradeHistory": [Function],
  "listBalances": [Function],
  "listOpenOrders": [Function],
  "listPositions": [Function],
  "logs": [
    {
      "msg": "Debug mode enabled",
    },
  ],
  "placeOrder": [Function],
  "placeOrders": [Function],
  "publicClient": {
    "account": undefined,
    "batch": undefined,
    "cacheTime": 4000,
    "call": [Function],
    "ccipRead": undefined,
    "chain": {
      "blockExplorers": {
        "default": {
          "apiUrl": "https://api-sepolia.arbiscan.io/api",
          "name": "Arbiscan",
          "url": "https://sepolia.arbiscan.io",
        },
      },
      "contracts": {
        "multicall3": {
          "address": "0xca11bde05977b3631167028862be2a173976ca11",
          "blockCreated": 81930,
        },
      },
      "fees": undefined,
      "formatters": undefined,
      "id": 421614,
      "name": "Arbitrum Sepolia",
      "nativeCurrency": {
        "decimals": 18,
        "name": "Arbitrum Sepolia Ether",
        "symbol": "ETH",
      },
      "rpcUrls": {
        "default": {
          "http": [
            "https://sepolia-rollup.arbitrum.io/rpc",
          ],
        },
      },
      "serializers": undefined,
      "testnet": true,
    },
    "createBlockFilter": [Function],
    "createContractEventFilter": [Function],
    "createEventFilter": [Function],
    "createPendingTransactionFilter": [Function],
    "estimateContractGas": [Function],
    "estimateFeesPerGas": [Function],
    "estimateGas": [Function],
    "estimateMaxPriorityFeePerGas": [Function],
    "extend": [Function],
    "getBalance": [Function],
    "getBlobBaseFee": [Function],
    "getBlock": [Function],
    "getBlockNumber": [Function],
    "getBlockTransactionCount": [Function],
    "getBytecode": [Function],
    "getChainId": [Function],
    "getCode": [Function],
    "getContractEvents": [Function],
    "getEip712Domain": [Function],
    "getEnsAddress": [Function],
    "getEnsAvatar": [Function],
    "getEnsName": [Function],
    "getEnsResolver": [Function],
    "getEnsText": [Function],
    "getFeeHistory": [Function],
    "getFilterChanges": [Function],
    "getFilterLogs": [Function],
    "getGasPrice": [Function],
    "getLogs": [Function],
    "getProof": [Function],
    "getStorageAt": [Function],
    "getTransaction": [Function],
    "getTransactionConfirmations": [Function],
    "getTransactionCount": [Function],
    "getTransactionReceipt": [Function],
    "key": "public",
    "multicall": [Function],
    "name": "Public Client",
    "pollingInterval": 4000,
    "prepareTransactionRequest": [Function],
    "readContract": [Function],
    "request": [Function],
    "sendRawTransaction": [Function],
    "simulateContract": [Function],
    "transport": {
      "fetchOptions": undefined,
      "key": "http",
      "name": "HTTP JSON-RPC",
      "request": [Function],
      "retryCount": 3,
      "retryDelay": 150,
      "timeout": 10000,
      "type": "http",
      "url": "https://test-rpc.quiknode.pro",
    },
    "type": "publicClient",
    "uid": "test-uid",
    "uninstallFilter": [Function],
    "verifyMessage": [Function],
    "verifySiweMessage": [Function],
    "verifyTypedData": [Function],
    "waitForTransactionReceipt": [Function],
    "watchBlockNumber": [Function],
    "watchBlocks": [Function],
    "watchContractEvent": [Function],
    "watchEvent": [Function],
    "watchPendingTransactions": [Function],
  },
  "rpc": "https://test-rpc.quiknode.pro",
  "subAccountId": 2,
  "verifierAddress": "0x27809a3Bd3cf44d855f1BE668bFD16D34bcE157C",
  "withdraw": [Function],
}
`;

exports[`The RyskSDK > should initialise correctly with only required params 1`] = `
RyskSDK {
  "account": {
    "address": "0xb47B0b1e44B932Ae9Bb01817E7010A553A965Ea8",
    "experimental_signAuthorization": [Function],
    "nonceManager": undefined,
    "publicKey": "0x0422eea015a8664523fce2eed5de514ad98a29724755c488f3d11d913d5671c881bc2d7b1e0d17389604ef141a96830b14917b034b7621579d1159d694dd231798",
    "sign": [Function],
    "signMessage": [Function],
    "signTransaction": [Function],
    "signTypedData": [Function],
    "source": "privateKey",
    "type": "local",
  },
  "calculateMarginRequirement": [Function],
  "cancelAndReplaceOrder": [Function],
  "cancelOpenOrdersForProduct": [Function],
  "cancelOrder": [Function],
  "cancelOrders": [Function],
  "chain": 421614,
  "ciaoAddress": "0x71728FDDF90233cc35D61bec7858d7c42A310ACe",
  "deposit": [Function],
  "domain": {
    "chainId": 421614n,
    "name": "rysk",
    "verifyingContract": "0x27809a3Bd3cf44d855f1BE668bFD16D34bcE157C",
    "version": "0.0.0",
  },
  "environment": "testnet",
  "getAccountHealth": [Function],
  "getKlines": [Function],
  "getOrderBook": [Function],
  "getProduct": [Function],
  "getProducts": [Function],
  "getServerTime": [Function],
  "getTickers": [Function],
  "getTradeHistory": [Function],
  "listBalances": [Function],
  "listOpenOrders": [Function],
  "listPositions": [Function],
  "logs": [],
  "placeOrder": [Function],
  "placeOrders": [Function],
  "publicClient": {
    "account": undefined,
    "batch": undefined,
    "cacheTime": 4000,
    "call": [Function],
    "ccipRead": undefined,
    "chain": {
      "blockExplorers": {
        "default": {
          "apiUrl": "https://api-sepolia.arbiscan.io/api",
          "name": "Arbiscan",
          "url": "https://sepolia.arbiscan.io",
        },
      },
      "contracts": {
        "multicall3": {
          "address": "0xca11bde05977b3631167028862be2a173976ca11",
          "blockCreated": 81930,
        },
      },
      "fees": undefined,
      "formatters": undefined,
      "id": 421614,
      "name": "Arbitrum Sepolia",
      "nativeCurrency": {
        "decimals": 18,
        "name": "Arbitrum Sepolia Ether",
        "symbol": "ETH",
      },
      "rpcUrls": {
        "default": {
          "http": [
            "https://sepolia-rollup.arbitrum.io/rpc",
          ],
        },
      },
      "serializers": undefined,
      "testnet": true,
    },
    "createBlockFilter": [Function],
    "createContractEventFilter": [Function],
    "createEventFilter": [Function],
    "createPendingTransactionFilter": [Function],
    "estimateContractGas": [Function],
    "estimateFeesPerGas": [Function],
    "estimateGas": [Function],
    "estimateMaxPriorityFeePerGas": [Function],
    "extend": [Function],
    "getBalance": [Function],
    "getBlobBaseFee": [Function],
    "getBlock": [Function],
    "getBlockNumber": [Function],
    "getBlockTransactionCount": [Function],
    "getBytecode": [Function],
    "getChainId": [Function],
    "getCode": [Function],
    "getContractEvents": [Function],
    "getEip712Domain": [Function],
    "getEnsAddress": [Function],
    "getEnsAvatar": [Function],
    "getEnsName": [Function],
    "getEnsResolver": [Function],
    "getEnsText": [Function],
    "getFeeHistory": [Function],
    "getFilterChanges": [Function],
    "getFilterLogs": [Function],
    "getGasPrice": [Function],
    "getLogs": [Function],
    "getProof": [Function],
    "getStorageAt": [Function],
    "getTransaction": [Function],
    "getTransactionConfirmations": [Function],
    "getTransactionCount": [Function],
    "getTransactionReceipt": [Function],
    "key": "public",
    "multicall": [Function],
    "name": "Public Client",
    "pollingInterval": 4000,
    "prepareTransactionRequest": [Function],
    "readContract": [Function],
    "request": [Function],
    "sendRawTransaction": [Function],
    "simulateContract": [Function],
    "transport": {
      "fetchOptions": undefined,
      "key": "http",
      "name": "HTTP JSON-RPC",
      "request": [Function],
      "retryCount": 3,
      "retryDelay": 150,
      "timeout": 10000,
      "type": "http",
      "url": "https://sepolia-rollup.arbitrum.io/rpc",
    },
    "type": "publicClient",
    "uid": "test-uid",
    "uninstallFilter": [Function],
    "verifyMessage": [Function],
    "verifySiweMessage": [Function],
    "verifyTypedData": [Function],
    "waitForTransactionReceipt": [Function],
    "watchBlockNumber": [Function],
    "watchBlocks": [Function],
    "watchContractEvent": [Function],
    "watchEvent": [Function],
    "watchPendingTransactions": [Function],
  },
  "rpc": "https://sepolia-rollup.arbitrum.io/rpc",
  "subAccountId": 1,
  "verifierAddress": "0x27809a3Bd3cf44d855f1BE668bFD16D34bcE157C",
  "withdraw": [Function],
}
`;
